src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+	gr->TD0tc  = gr->grAttr.d0Typecode;
+	gr->TD1tc  = gr->grAttr.d1Typecode;
+	gr->TI0tc  = gr->grAttr.i0Typecode;
+	gr->TS32tc = GA_INT;
+	gr->TU32tc = GA_UINT;
+	gr->TS64tc = GA_LONG;
+	gr->TU64tc = GA_ULONG;
+	reduxGenSetKTypes(gr);
+	
+	
+	/**
+	 * Compute number of kernel arguments and construct kernel argument
+	 * typecode list.
+	 */
+	
+	reduxGenSetMaxBS(gr);
+	reduxGenIterArgs(gr, reduxGenCountArgs, &gr->kNumArgs);
+	gr->kArgTypeCodes = calloc(gr->kNumArgs, sizeof(*gr->kArgTypeCodes));
+	if (!gr->kArgTypeCodes){
+		return reduxGenCleanupMsg(gr, GA_MEMORY_ERROR,
+		                          "Failed to allocate memory for kernel arguments "
+		                          "typecode list!\n");
+	}
+	i = 0;
+	reduxGenIterArgs(gr, reduxGenSaveArgTypecodes, &i);
+	
+	
+	/* Generate source code. */
+	return reduxGenSrc(gr);
+}
+
+/**
+ * Compute maximum block size we shall support in generated kernels.
+ */
+
+static void        reduxGenSetMaxBS                 (GpuReduction*        gr){
+	gr->maxBS = gr->grAttr.maxLM/reduxGenGetReduxStateSize(gr);
+	gr->maxBS = gr->maxBS < gr->grAttr.maxLg ? gr->maxBS : gr->grAttr.maxLg;
+	gr->maxBS = gr->maxBS < gr->grAttr.maxL0 ? gr->maxBS : gr->grAttr.maxL0;
+	
+	/**
+	 * In practice we want a moderate amount of blocks, not just one monolith
+	 * that occupies a processor for its entire lifetime. E.g. An NVIDIA GPU
+	 * supports 1024 threads / block, but we shall gun for less than that.
+	 * 
+	 * Our heuristic shall be to divide by 4 the maximum number of threads per
+	 * block, so that there's 4 times more blocks than normally there would be.
+	 * This helps on many fronts:
+	 * 
+	 *   - A smaller "tail effect" when the last huge block must wait its turn
+	 *     and then delays the completion of the entire grid
+	 *   - The horizontal reductions take less time per block, and sometimes
+	 *     horizontal reduction time can dominate performance.
+	 *   - Less time taken for across-thread synchronization; And whenever a
+	 *     block's threads are stalled waiting for synchronization, another
+	 *     block's threads can fill in with their global memory requests.
+	 */
+	
+	if (gr->maxBS >= 16){
+		gr->maxBS /= 4;
+	}
+	
+	/* Since ceil(log2(maxBS)) is also heavily used, compute it here */
+	gr->log2MaxBS = reduxCeilLog2(gr->maxBS);
+}
+
+/**
+ * Decide on the TK* accumulator types and initializers we will use.
+ * 
+ * Currently, the only special thing we do is to promote the accumulator type
+ * to GA_FLOATx if the source type is GA_HALFx:
+ * 
+ *     TPS0 = promotion(TS0)
+ * 
+ * Therefore, it is currently always the case that TK0 == TPS0.
+ * 
+ * In the future this might become wierder when the accumulator is a Kahan
+ * summation, for instance, and then TK0 != promoted(TS0).
+ * 
+ * If the user guaranteed to us through gr->grAttr that TK1 can be made
+ * narrower than 64-bit, this is also where we'd take this into account.
+ * For now we default TK1 to exactly TI0.
```

Comment:
@abergeron See above comment, but also note that for this kernel I personally assume responsibility for selecting 32/64-bit-length code versions. It's what the `selector` does - It selects one of 8 possible "templates" for Phase 0, or it selects the one version of the fixup code in Phase 1.